{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5100e487",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading swap table from: /Users/ishaharris/Projects/ribolings/data/CodonSwapTable_withStop.csv\n",
      "Swap table loaded: 64 rows, 21 amino acids found\n",
      "Building AA-normalised nucleotide frequencies (draws per AA = 100)\n",
      "Built AA freq table for 21 AAs\n",
      "Found 25 FASTA files in /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta\n",
      "Processing 11 selected FASTA files\n",
      "[1/11] Processing 'Arena' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Arena.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/ribolings/lib/python3.11/site-packages/Bio/Seq.py:2334: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Seqs: 249, skipped_nonstandard: 12, skipped_no_prediction: 0\n",
      "[2/11] Processing 'Calici' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Calici.fasta\n",
      "  Seqs: 72, skipped_nonstandard: 3, skipped_no_prediction: 0\n",
      "[3/11] Processing 'Corona' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Corona.fasta\n",
      "  Seqs: 458, skipped_nonstandard: 11, skipped_no_prediction: 0\n",
      "[4/11] Processing 'Filo' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Filo.fasta\n",
      "  Seqs: 125, skipped_nonstandard: 2, skipped_no_prediction: 0\n",
      "[5/11] Processing 'Flavi' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Flavi.fasta\n",
      "  Seqs: 60, skipped_nonstandard: 14, skipped_no_prediction: 0\n",
      "[6/11] Processing 'Hanta' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Hanta.fasta\n",
      "  Seqs: 131, skipped_nonstandard: 5, skipped_no_prediction: 0\n",
      "[7/11] Processing 'Orthomyxo' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Orthomyxo.fasta\n",
      "  Seqs: 67, skipped_nonstandard: 1, skipped_no_prediction: 0\n",
      "[8/11] Processing 'Paramyxo' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Paramyxo.fasta\n",
      "  Seqs: 647, skipped_nonstandard: 5, skipped_no_prediction: 0\n",
      "[9/11] Processing 'Peribunya' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Peribunya.fasta\n",
      "  Seqs: 3, skipped_nonstandard: 0, skipped_no_prediction: 0\n",
      "[10/11] Processing 'Picorna' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Picorna.fasta\n",
      "  Seqs: 648, skipped_nonstandard: 94, skipped_no_prediction: 0\n",
      "[11/11] Processing 'Rhabdo' -> /Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta/Rhabdo.fasta\n",
      "  Seqs: 193, skipped_nonstandard: 5, skipped_no_prediction: 0\n"
     ]
    },
    {
     "ename": "IsADirectoryError",
     "evalue": "[Errno 21] Is a directory: '/Users/ishaharris/Projects/ribolings/data/freqs'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mIsADirectoryError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 192\u001b[39m\n\u001b[32m    180\u001b[39m         results.append({\n\u001b[32m    181\u001b[39m             \u001b[33m'\u001b[39m\u001b[33mVirusFamily\u001b[39m\u001b[33m'\u001b[39m: family,\n\u001b[32m    182\u001b[39m             \u001b[33m'\u001b[39m\u001b[33mA_pred\u001b[39m\u001b[33m'\u001b[39m: stats[\u001b[33m'\u001b[39m\u001b[33mA_pred\u001b[39m\u001b[33m'\u001b[39m],\n\u001b[32m   (...)\u001b[39m\u001b[32m    188\u001b[39m             \u001b[33m'\u001b[39m\u001b[33mn_skipped_no_prediction\u001b[39m\u001b[33m'\u001b[39m: stats[\u001b[33m'\u001b[39m\u001b[33mn_skipped_no_prediction\u001b[39m\u001b[33m'\u001b[39m],\n\u001b[32m    189\u001b[39m         })\n\u001b[32m    191\u001b[39m out_df = pd.DataFrame(results)\n\u001b[32m--> \u001b[39m\u001b[32m192\u001b[39m \u001b[43mout_df\u001b[49m\u001b[43m.\u001b[49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mOUTPUT_CSV\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m    193\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mSaved predicted nucleotide frequencies for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(out_df)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m FASTA files to \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mOUTPUT_CSV\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m    194\u001b[39m out_df\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/ribolings/lib/python3.11/site-packages/pandas/util/_decorators.py:333\u001b[39m, in \u001b[36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    327\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) > num_allow_args:\n\u001b[32m    328\u001b[39m     warnings.warn(\n\u001b[32m    329\u001b[39m         msg.format(arguments=_format_argument_list(allow_args)),\n\u001b[32m    330\u001b[39m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[32m    331\u001b[39m         stacklevel=find_stack_level(),\n\u001b[32m    332\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m333\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/ribolings/lib/python3.11/site-packages/pandas/core/generic.py:3986\u001b[39m, in \u001b[36mNDFrame.to_csv\u001b[39m\u001b[34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[39m\n\u001b[32m   3975\u001b[39m df = \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m.to_frame()\n\u001b[32m   3977\u001b[39m formatter = DataFrameFormatter(\n\u001b[32m   3978\u001b[39m     frame=df,\n\u001b[32m   3979\u001b[39m     header=header,\n\u001b[32m   (...)\u001b[39m\u001b[32m   3983\u001b[39m     decimal=decimal,\n\u001b[32m   3984\u001b[39m )\n\u001b[32m-> \u001b[39m\u001b[32m3986\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   3987\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3988\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3989\u001b[39m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3990\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3991\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3992\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3993\u001b[39m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[43m=\u001b[49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3994\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3995\u001b[39m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[43m=\u001b[49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3996\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3997\u001b[39m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m=\u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3998\u001b[39m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[43m=\u001b[49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3999\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4000\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4001\u001b[39m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[43m=\u001b[49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4002\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4003\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/ribolings/lib/python3.11/site-packages/pandas/io/formats/format.py:1014\u001b[39m, in \u001b[36mDataFrameRenderer.to_csv\u001b[39m\u001b[34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[39m\n\u001b[32m    993\u001b[39m     created_buffer = \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[32m    995\u001b[39m csv_formatter = CSVFormatter(\n\u001b[32m    996\u001b[39m     path_or_buf=path_or_buf,\n\u001b[32m    997\u001b[39m     lineterminator=lineterminator,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1012\u001b[39m     formatter=\u001b[38;5;28mself\u001b[39m.fmt,\n\u001b[32m   1013\u001b[39m )\n\u001b[32m-> \u001b[39m\u001b[32m1014\u001b[39m \u001b[43mcsv_formatter\u001b[49m\u001b[43m.\u001b[49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1016\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[32m   1017\u001b[39m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/ribolings/lib/python3.11/site-packages/pandas/io/formats/csvs.py:251\u001b[39m, in \u001b[36mCSVFormatter.save\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    247\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    248\u001b[39m \u001b[33;03mCreate the writer & save.\u001b[39;00m\n\u001b[32m    249\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    250\u001b[39m \u001b[38;5;66;03m# apply compression and byte/text conversion\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m251\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    252\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    253\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    254\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    255\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    256\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    257\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    258\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m handles:\n\u001b[32m    259\u001b[39m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[32m    260\u001b[39m     \u001b[38;5;28mself\u001b[39m.writer = csvlib.writer(\n\u001b[32m    261\u001b[39m         handles.handle,\n\u001b[32m    262\u001b[39m         lineterminator=\u001b[38;5;28mself\u001b[39m.lineterminator,\n\u001b[32m   (...)\u001b[39m\u001b[32m    267\u001b[39m         quotechar=\u001b[38;5;28mself\u001b[39m.quotechar,\n\u001b[32m    268\u001b[39m     )\n\u001b[32m    270\u001b[39m     \u001b[38;5;28mself\u001b[39m._save()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/ribolings/lib/python3.11/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    868\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    869\u001b[39m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[32m    870\u001b[39m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[32m    871\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m         handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[32m    882\u001b[39m         handle = \u001b[38;5;28mopen\u001b[39m(handle, ioargs.mode)\n",
      "\u001b[31mIsADirectoryError\u001b[39m: [Errno 21] Is a directory: '/Users/ishaharris/Projects/ribolings/data/freqs'"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "AA Enrich Predicted Frequencies (Jupyter version)\n",
    "\n",
    "- Configure paths in the config section.\n",
    "- Set SELECT_FAMILIES to a list of virus-family basenames (no extension) to process only those,\n",
    "  or to None to process all FASTA files in FASTA_DIR.\n",
    "- Prints show progress.\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import random\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# -----------------------------\n",
    "# Config (edit these paths)\n",
    "# -----------------------------\n",
    "FASTA_DIR = \"/Users/ishaharris/Projects/ribolings/data/virus/rna_cds_fasta\"            # folder containing one FASTA per virus family\n",
    "SWAP_TABLE_CSV = \"/Users/ishaharris/Projects/ribolings/data/CodonSwapTable_withStop.csv\"\n",
    "OUTPUT_CSV = \"/Users/ishaharris/Projects/ribolings/data/freqs/monoaa_freqs.csv\"`\n",
    "RANDOM_SEED = 42\n",
    "N_DRAWS_PER_AA = 100\n",
    "\n",
    "# Set SELECT_FAMILIES to a list of basenames (filename without extension) to process only those.\n",
    "# Example: SELECT_FAMILIES = ['coronavirus', 'orthomyxovirus']\n",
    "# Or set to None to process every fasta in FASTA_DIR.\n",
    "SELECT_FAMILIES = ['Arena', 'Calici', 'Corona', 'Filo', 'Flavi', 'Hanta', 'Picorna', 'Orthomyxo', 'Paramyxo', 'Peribunya','Rhabdo']  \n",
    "\n",
    "random.seed(RANDOM_SEED)\n",
    "np.random.seed(RANDOM_SEED)\n",
    "\n",
    "# -----------------------------\n",
    "# Helpers\n",
    "# -----------------------------\n",
    "\n",
    "def load_swap_table(path):\n",
    "    print(f\"Loading swap table from: {path}\")\n",
    "    df = pd.read_csv(path)\n",
    "    cols = {c.lower(): c for c in df.columns}\n",
    "    required = ['codon', 'number', 'aa']\n",
    "    for r in required:\n",
    "        if r not in cols:\n",
    "            raise ValueError(f\"Swap table missing required column: {r}\")\n",
    "    df = df.rename(columns={cols['codon']:'Codon', cols['number']:'Number', cols['aa']:'AA'})\n",
    "    df['Codon'] = df['Codon'].astype(str).str.lower()\n",
    "    df['AA'] = df['AA'].astype(str)\n",
    "    print(f\"Swap table loaded: {len(df)} rows, {df['AA'].nunique()} amino acids found\")\n",
    "    return df\n",
    "\n",
    "\n",
    "def build_aa_freqs_normalised(swap_df, draws=N_DRAWS_PER_AA):\n",
    "    print(f\"Building AA-normalised nucleotide frequencies (draws per AA = {draws})\")\n",
    "    aa_groups = swap_df.groupby('AA')\n",
    "    aa_rows = {}\n",
    "    for aa, group in aa_groups:\n",
    "        codons = group['Codon'].tolist()\n",
    "        if not codons:\n",
    "            continue\n",
    "        drawn = [random.choice(codons) for _ in range(draws)]\n",
    "        concat = ''.join(drawn)\n",
    "        counts = Counter(list(concat))\n",
    "        total = len(concat)\n",
    "        freqs = {\n",
    "            'A': counts.get('a', 0) / total,\n",
    "            'C': counts.get('c', 0) / total,\n",
    "            'G': counts.get('g', 0) / total,\n",
    "            'T': counts.get('t', 0) / total,\n",
    "        }\n",
    "        aa_rows[aa] = {k: v * 3.0 for k, v in freqs.items()}\n",
    "    aa_freqs_norm = pd.DataFrame.from_dict(aa_rows, orient='index')\n",
    "    aa_freqs_norm = aa_freqs_norm[['A','C','G','T']]\n",
    "    print(f\"Built AA freq table for {len(aa_freqs_norm)} AAs\")\n",
    "    return aa_freqs_norm\n",
    "\n",
    "\n",
    "def predict_nt_freq_for_seq(record_seq, aa_freqs_norm):\n",
    "    seq = Seq(str(record_seq))\n",
    "    translation = seq.translate()\n",
    "    aa_list = list(str(translation))\n",
    "    if not aa_list:\n",
    "        return None\n",
    "    nt_sums = {'A': 0.0, 'C': 0.0, 'G': 0.0, 'T': 0.0}\n",
    "    used_aas = 0\n",
    "    for aa in aa_list:\n",
    "        aa_key = 'X' if aa == '*' else aa\n",
    "        if aa_key not in aa_freqs_norm.index:\n",
    "            continue\n",
    "        row = aa_freqs_norm.loc[aa_key]\n",
    "        for nt in ['A','C','G','T']:\n",
    "            nt_sums[nt] += float(row[nt])\n",
    "        used_aas += 1\n",
    "    if used_aas == 0:\n",
    "        return None\n",
    "    total_nt = used_aas * 3.0\n",
    "    return {nt: nt_sums[nt] / total_nt for nt in ['A','C','G','T']}\n",
    "\n",
    "\n",
    "def seq_is_standard_dna(seq_str):\n",
    "    return set(seq_str.lower()).issubset({'a','c','g','t'})\n",
    "\n",
    "\n",
    "def average_predicted_freqs_for_fasta(fasta_path, aa_freqs_norm):\n",
    "    predicted_list = []\n",
    "    n_total, n_skipped_nonstandard, n_skipped_no_prediction = 0, 0, 0\n",
    "    for rec in SeqIO.parse(fasta_path, 'fasta'):\n",
    "        n_total += 1\n",
    "        seq_str = str(rec.seq)\n",
    "        if not seq_is_standard_dna(seq_str):\n",
    "            n_skipped_nonstandard += 1\n",
    "            continue\n",
    "        pred = predict_nt_freq_for_seq(seq_str, aa_freqs_norm)\n",
    "        if pred is None:\n",
    "            n_skipped_no_prediction += 1\n",
    "            continue\n",
    "        predicted_list.append([pred['A'], pred['C'], pred['G'], pred['T']])\n",
    "    if not predicted_list:\n",
    "        return {'A_pred': np.nan,'C_pred': np.nan,'G_pred': np.nan,'T_pred': np.nan,\n",
    "                'n_seqs': n_total,'n_skipped_nonstandard': n_skipped_nonstandard,\n",
    "                'n_skipped_no_prediction': n_skipped_no_prediction}\n",
    "    arr = np.array(predicted_list)\n",
    "    mean_vals = np.nanmean(arr, axis=0)\n",
    "    return {\n",
    "        'A_pred': float(mean_vals[0]),\n",
    "        'C_pred': float(mean_vals[1]),\n",
    "        'G_pred': float(mean_vals[2]),\n",
    "        'T_pred': float(mean_vals[3]),\n",
    "        'n_seqs': n_total,\n",
    "        'n_skipped_nonstandard': n_skipped_nonstandard,\n",
    "        'n_skipped_no_prediction': n_skipped_no_prediction,\n",
    "    }\n",
    "\n",
    "# -----------------------------\n",
    "# Run in notebook cell\n",
    "# -----------------------------\n",
    "\n",
    "# Load swap table and build AA freq table\n",
    "swap_df = load_swap_table(SWAP_TABLE_CSV)\n",
    "aa_freqs_norm = build_aa_freqs_normalised(swap_df, draws=N_DRAWS_PER_AA)\n",
    "if '*' in aa_freqs_norm.index and 'X' not in aa_freqs_norm.index:\n",
    "    aa_freqs_norm = aa_freqs_norm.rename(index={'*':'X'})\n",
    "\n",
    "# List FASTA files\n",
    "all_fnames = sorted([f for f in os.listdir(FASTA_DIR) if f.lower().endswith(('.fa','.fasta','.fas'))])\n",
    "print(f\"Found {len(all_fnames)} FASTA files in {FASTA_DIR}\")\n",
    "\n",
    "# Prepare list of files to process based on SELECT_FAMILIES\n",
    "if SELECT_FAMILIES is None:\n",
    "    to_process = all_fnames\n",
    "    print(\"SELECT_FAMILIES is None -> processing all FASTA files\")\n",
    "else:\n",
    "    sel_set = set(SELECT_FAMILIES)\n",
    "    to_process = []\n",
    "    for f in all_fnames:\n",
    "        base = os.path.splitext(f)[0]\n",
    "        if base in sel_set:\n",
    "            to_process.append(f)\n",
    "    missing = sel_set - set(os.path.splitext(f)[0] for f in all_fnames)\n",
    "    if missing:\n",
    "        print(f\"Warning: the following selected families were not found in {FASTA_DIR}: {sorted(missing)}\")\n",
    "    print(f\"Processing {len(to_process)} selected FASTA files\")\n",
    "\n",
    "results = []\n",
    "if not to_process:\n",
    "    print(\"No FASTA files to process. Exiting.\")\n",
    "else:\n",
    "    for idx, fname in enumerate(to_process, start=1):\n",
    "        path = os.path.join(FASTA_DIR, fname)\n",
    "        family = os.path.splitext(fname)[0]\n",
    "        print(f\"[{idx}/{len(to_process)}] Processing '{family}' -> {path}\")\n",
    "        try:\n",
    "            stats = average_predicted_freqs_for_fasta(path, aa_freqs_norm)\n",
    "        except Exception as e:\n",
    "            print(f\"  Error processing {fname}: {e}\")\n",
    "            stats = {'A_pred': np.nan,'C_pred': np.nan,'G_pred': np.nan,'T_pred': np.nan,\n",
    "                     'n_seqs': 0,'n_skipped_nonstandard': 0,'n_skipped_no_prediction': 0}\n",
    "        print(f\"  Seqs: {stats['n_seqs']}, skipped_nonstandard: {stats['n_skipped_nonstandard']}, skipped_no_prediction: {stats['n_skipped_no_prediction']}\")\n",
    "        results.append({\n",
    "            'VirusFamily': family,\n",
    "            'A_pred': stats['A_pred'],\n",
    "            'C_pred': stats['C_pred'],\n",
    "            'G_pred': stats['G_pred'],\n",
    "            'T_pred': stats['T_pred'],\n",
    "            'n_seqs': stats['n_seqs'],\n",
    "            'n_skipped_nonstandard': stats['n_skipped_nonstandard'],\n",
    "            'n_skipped_no_prediction': stats['n_skipped_no_prediction'],\n",
    "        })\n",
    "\n",
    "out_df = pd.DataFrame(results)\n",
    "out_df.to_csv(OUTPUT_CSV, index=False)\n",
    "print(f\"Saved predicted nucleotide frequencies for {len(out_df)} FASTA files to {OUTPUT_CSV}\")\n",
    "out_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bbfab8ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved predicted nucleotide frequencies for 11 FASTA files to /Users/ishaharris/Projects/ribolings/data/freqs/monoaa_freqs.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VirusFamily</th>\n",
       "      <th>A_pred</th>\n",
       "      <th>C_pred</th>\n",
       "      <th>G_pred</th>\n",
       "      <th>T_pred</th>\n",
       "      <th>n_seqs</th>\n",
       "      <th>n_skipped_nonstandard</th>\n",
       "      <th>n_skipped_no_prediction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Arena</td>\n",
       "      <td>0.286640</td>\n",
       "      <td>0.227732</td>\n",
       "      <td>0.234334</td>\n",
       "      <td>0.251294</td>\n",
       "      <td>249</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Calici</td>\n",
       "      <td>0.251726</td>\n",
       "      <td>0.257963</td>\n",
       "      <td>0.259127</td>\n",
       "      <td>0.231184</td>\n",
       "      <td>72</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Corona</td>\n",
       "      <td>0.261607</td>\n",
       "      <td>0.232161</td>\n",
       "      <td>0.230950</td>\n",
       "      <td>0.275283</td>\n",
       "      <td>458</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Filo</td>\n",
       "      <td>0.270840</td>\n",
       "      <td>0.250647</td>\n",
       "      <td>0.242735</td>\n",
       "      <td>0.235778</td>\n",
       "      <td>125</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Flavi</td>\n",
       "      <td>0.240566</td>\n",
       "      <td>0.244986</td>\n",
       "      <td>0.274156</td>\n",
       "      <td>0.240293</td>\n",
       "      <td>60</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hanta</td>\n",
       "      <td>0.282991</td>\n",
       "      <td>0.224665</td>\n",
       "      <td>0.246442</td>\n",
       "      <td>0.245902</td>\n",
       "      <td>131</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Orthomyxo</td>\n",
       "      <td>0.291676</td>\n",
       "      <td>0.212279</td>\n",
       "      <td>0.265402</td>\n",
       "      <td>0.230643</td>\n",
       "      <td>67</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Paramyxo</td>\n",
       "      <td>0.287148</td>\n",
       "      <td>0.236739</td>\n",
       "      <td>0.242440</td>\n",
       "      <td>0.233673</td>\n",
       "      <td>647</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Peribunya</td>\n",
       "      <td>0.285585</td>\n",
       "      <td>0.197754</td>\n",
       "      <td>0.287368</td>\n",
       "      <td>0.229294</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Picorna</td>\n",
       "      <td>0.273023</td>\n",
       "      <td>0.241535</td>\n",
       "      <td>0.241724</td>\n",
       "      <td>0.243717</td>\n",
       "      <td>648</td>\n",
       "      <td>94</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Rhabdo</td>\n",
       "      <td>0.281819</td>\n",
       "      <td>0.226970</td>\n",
       "      <td>0.252167</td>\n",
       "      <td>0.239043</td>\n",
       "      <td>193</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   VirusFamily    A_pred    C_pred    G_pred    T_pred  n_seqs  \\\n",
       "0        Arena  0.286640  0.227732  0.234334  0.251294     249   \n",
       "1       Calici  0.251726  0.257963  0.259127  0.231184      72   \n",
       "2       Corona  0.261607  0.232161  0.230950  0.275283     458   \n",
       "3         Filo  0.270840  0.250647  0.242735  0.235778     125   \n",
       "4        Flavi  0.240566  0.244986  0.274156  0.240293      60   \n",
       "5        Hanta  0.282991  0.224665  0.246442  0.245902     131   \n",
       "6    Orthomyxo  0.291676  0.212279  0.265402  0.230643      67   \n",
       "7     Paramyxo  0.287148  0.236739  0.242440  0.233673     647   \n",
       "8    Peribunya  0.285585  0.197754  0.287368  0.229294       3   \n",
       "9      Picorna  0.273023  0.241535  0.241724  0.243717     648   \n",
       "10      Rhabdo  0.281819  0.226970  0.252167  0.239043     193   \n",
       "\n",
       "    n_skipped_nonstandard  n_skipped_no_prediction  \n",
       "0                      12                        0  \n",
       "1                       3                        0  \n",
       "2                      11                        0  \n",
       "3                       2                        0  \n",
       "4                      14                        0  \n",
       "5                       5                        0  \n",
       "6                       1                        0  \n",
       "7                       5                        0  \n",
       "8                       0                        0  \n",
       "9                      94                        0  \n",
       "10                      5                        0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OUTPUT_CSV = \"/Users/ishaharris/Projects/ribolings/data/freqs/monoaa_freqs.csv\"\n",
    "\n",
    "out_df = pd.DataFrame(results)\n",
    "out_df.to_csv(OUTPUT_CSV, index=False)\n",
    "print(f\"Saved predicted nucleotide frequencies for {len(out_df)} FASTA files to {OUTPUT_CSV}\")\n",
    "out_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ribolings",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
